docs/EVALUATION.md

```diff
@@ -0,0 +1,249 @@
+
+##### More fine grained view
+
+Accounting for the afore mentioned assumptions that training takes some time and triggers are not at one point in time but specify a training batch (time interval), we can visualize a pipeline run as follows:
+
+
+```mermaid
+gantt
+    title A Gantt Diagram
+    dateFormat YYYY-MM-DD
+    axisFormat %Y
+    section models
+        Trigger1 :milestone, t1, 1930, 0d
+        train_batch_1 :a10, 1928, 2y
+        training1 :a11, after t1, 1y
+        model 1          :a12, after a11, until a22
+
+        Trigger2 :milestone, t2, 1940, 0d
+        train_batch_2 :a20, after a10, until t2
+        training2 :a21, after t2, 1y
+        model 2          :a22, after a21, until a32
+
+        Trigger3 :milestone, t3, 1945, 0d
+        train_batch_3 :a30, after a20, until t3
+        training3 :a31, after t3, 1y
+        model 3          :a32, after a31, until a42
+
+        Trigger4 :milestone, t4, 1950, 0d
+        train_batch_4 :a40, after a30, until t4
+        training4 :a41, after t4, 1y
+        model 4          :a42, after a41, 10y
+
+```
+
+With our assumption of `training_time=0` this `training_<i>` would be can be removed so that we can evaluate one model on the data between two triggers (hence the next training batch)
+
+##### State of modyn (at the time of writing)
+
+We currectly evaluate a model directly after training it (e.g. we are now at `Trigger 4`) and have now robust approach of combining the results of the single model evaluations to a pipeline performance. We support a `OfflineEvaluationStrategy` that specifies the interval in which we want to evaluate the models. We also support `MatrixEvaluationStrategy`: We evaluate on predefined grid of evaluation points.
+
+We face several problems:
+
+1) We don't know when the next trigger occurs.
+    This might be fine if we assume a time based trigger policy with a regular schedule where we can set the OfflineEvaluationStrategy lookahead offset to the trigger interval. However, for a generic trigger policy we cannot assume this. In consequence when setting a fixed lookahead offset we might end up evaluating a model with a lookahead offset that is too large or too small. (E.g. when always evaluate on the next 5 years after a training. That would mean that if a policy doesn't retrain for 20 years we miss a lot of data)
+
```

Comment:
But evaluation is not necessarily coupled to the training: still every once in a while you can do evaluation on the newly arrived data, right?

Reply:
Conceptually, yes, implementing that is probably rather tricky with the current assumptions in the evaluation logic. Overall that describes the approach of pipeline feedback loop for triggering quite well.

Reply:
What's the assumption that lets implementing this tricky? I think we just need to refactor "where to evaluate", and uses an `OffsetEvalStrategy` to evaluate on the past predefined window of data.